README.md
<div align="center">
	<img src="https://github.com/alumna/pulsa/raw/master/pulsa.svg?sanitize=true" alt="pulsa" width="480" height="270" />
</div>

<div align="center">
	<a href="https://npmjs.org/package/@alumna/pulsa">
		<img src="https://badgen.now.sh/npm/v/@alumna/pulsa" alt="version" />
	</a>
	<a href="https://npmjs.org/package/@alumna/pulsa">
		<img src="https://badgen.net/bundlephobia/min/@alumna/pulsa" alt="size" />
	</a>
	<a href="https://travis-ci.org/alumna/pulsa">
		<img src="https://travis-ci.org/alumna/pulsa.svg?branch=master" alt="travis" />
	</a>
	<a href='https://coveralls.io/github/alumna/pulsa?branch=master'>
		<img src='https://coveralls.io/repos/github/alumna/pulsa/badge.svg?branch=master' alt='Coverage Status' />
	</a>
	<a href="https://npmjs.org/package/@alumna/pulsa">
		<img src="https://badgen.now.sh/npm/dm/@alumna/pulsa" alt="downloads" />
	</a>
</div>

<div align="center">In-memory static file server and middleware in Node</div>

<br/>

## Features

* Production version with **no dependencies**
* Extremely lightweight - 13kB!
* Works on any Express compatible framework, including [Polka](https://github.com/lukeed/polka)
* 100% tested and working on Linux, Mac and Windows!

Additionally, this module is delivered as:

* **ES Module**: [`dist/pulsa.es.js`](https://unpkg.com/@alumna/pulsa/dist/pulsa.es.js)
* **CommonJS**: [`dist/pulsa.cjs.js`](https://unpkg.com/@alumna/pulsa/dist/pulsa.cjs.js)


## Install

```
$ npm install @alumna/pulsa
```


## Usage

```js
import http       from 'http';
import * as pulsa from '@alumna/pulsa';

http.createServer( pulsa({

	// (OPTIONAL)
	// The current project's directory '.' will be used if `dir` isn't passed
	dir: './public_html/',

	// (OPTIONAL)
	// Mask paths to a different directory when requested
	alias: {
		'/old': '/new'
	},

	// (OPTIONAL)
	// Run as a SPA, mapping non-existent URL's to [base]index.html
	// Default to "false"
	spa: false,

	// (OPTIONAL)
	// Base directory to be considered running as a SPA
	// Default to '/'
	base: '/',

	// (OPTIONAL)
	// Define the max size (in bytes) of a file to be cached
	// Files greater than this limit will be streamed from disk
	// Default to "1048576" (which is 1MB)
	maxFileSize: 1048576
	
});

// Create a memory-file on the server
// "content" can be a string or a buffer
// -
// It will be instantly available to all instances where the path matches
// Works with SPA, index.html files served on dirs, updates 404 caches, etc
pulsa.memory( fullpath, content )

// Clear the cache of a file
// MUST be a full path, not relative.
// -
// It will instantly reflect the operation on all instances
pulsa.clear( '/var/www/index.html' )
```


CHANGELOG.md
# Alumna changelog

## 1.0.5 - 2020-07-21

* Named exports
* Adding Coveralls as a fallback to Codecov
* Updated devDependencies

## 1.0.4 - 2020-03-12

* Code quality fixes
* Updated devDependencies (rollup)

## 1.0.3 - 2020-03-11

* Updated devDependencies

## 1.0.2 - 2019-07-18

* Fix package.json entrypoint

## 1.0.1 - 2019-07-18

* Reduce complexity of most part of the functions
* Path adjustment to solve Windows support

## 1.0.0 - 2019-07-18

* First release of Pulsa

test/pulsa.test.js
import * as pulsa 	from './../src/index';
import fs 			from 'fs';
import http 		from 'http';
import { resolve }	from 'path';

const next = ( res ) => {

	return function () {
		res.statusCode = 200;
		res.write( 'External middlewares are being called on 404' );
		res.end();
	}

}



describe('Pulsa Server tests', () => {

	beforeAll(() => {
		http.createServer( pulsa.serve( { dir: './test/public_html' } ) ).listen( 9000 )
		http.createServer( pulsa.serve( './test/public_html' ) ).listen( 9001 )
		http.createServer( pulsa.serve( { dir: './test/public_spa', spa: true, base: '/spa' } ) ).listen( 9002 )
		http.createServer( pulsa.serve( { dir: './test/public_max', maxFileSize: 5 } ) ).listen( 9003 )
		http.createServer( function ( req, res ) { pulsa.serve( './test/public_html' )( req, res, next( res ) ) } ).listen( 9004 )

		http.createServer( pulsa.serve( {
			dir: './test/public_alias',
			alias: {
				'/components' : '/compiled'
			}
		})).listen( 9005 )

		http.createServer( pulsa.serve( { dir: './test/public_changing', spa: true } ) ).listen( 9006 )
		http.createServer( pulsa.serve( { dir: './test/public_spa_memory', spa: true } ) ).listen( 9007 )
		http.createServer( pulsa.serve( { dir: './test/public_memory', spa: false } ) ).listen( 9008 )
		http.createServer( pulsa.serve( { dir: './test/public_overwriting', spa: false } ) ).listen( 9009 )

		return;
	});

	test('1. Serving a HTML file with "Hello world!"', done => {

		http.get( 'http://127.0.0.1:9000', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'Hello world!' );
				done();

			})

		})

	});

	test('2. 404 on non-existent files"', done => {

		http.get( 'http://127.0.0.1:9000/this_doesnt_exist.txt', response => {

			expect( response.statusCode ).toBe( 404 );
			done();

		})

	});

	test('3. 404 on directories without files"', done => {

		http.get( 'http://127.0.0.1:9000/without_index', response => {

			expect( response.statusCode ).toBe( 404 );
			done();

		})

	});

	test('4. Automatically serving index.html inside a directory"', done => {

		const request = ( path = '', end = false ) => {

			http.get( 'http://127.0.0.1:9000' + path, response => {

				expect( response.statusCode ).toBe( 200 );

				let data = '';

				response.on( 'data', chunk => { data += chunk } )

				response.on( 'end', () => {

					expect( data ).toBe( 'Hello index!' );
					
					if ( end ) done();

				})

			})

		}

		// One request to '/with_index/index.html' to cover the line
		// that check the cache of this index.html on "stats"
		request( '/with_index/index.html' )

		// And one request to '/', to trigger the stats cache
		request( '/with_index/', true )

	});

	test('5. Serving a SPA', done => {

		const request = ( path = '', expected, end = false ) => {

			http.get( 'http://127.0.0.1:9002' + path, response => {

				expect( response.statusCode ).toBe( 200 );

				let data = '';

				response.on( 'data', chunk => { data += chunk } )

				response.on( 'end', () => {

					expect( data ).toBe( expected );
					
					if ( end ) done();

				})

			})

		}

		request( '/spa/', 'Hello SPA!' )
		request( '/spa/without_index', 'Hello SPA!' )
		request( '/spa/non_existent_dir', 'Hello SPA!' )
		request( '/spa/non_existent_file.txt', 'Hello SPA!' )
		request( '/spa/with_index_as_folder', 'Hello SPA!' )
		request( '/spa/with_index', 'Hello index inside a SPA!', true )

	});

	test('6. Serving range requests', done => {

		const options = {
			headers: { 'Range': 'bytes=0-4' },
		} 

		http.get( 'http://127.0.0.1:9000', options, response => {

			expect( response.statusCode ).toBe( 206 );
			expect( response.headers[ 'content-range' ] ).toBe( 'bytes 0-4/12' );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'Hello' );

			})

		})

		//

		options.headers[ 'Range' ] = 'bytes=6-'

		http.get( 'http://127.0.0.1:9000', options, response => {

			expect( response.statusCode ).toBe( 206 );
			expect( response.headers[ 'content-range' ] ).toBe( 'bytes 6-11/12' );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'world!' );
				done();

			})

		})

	});

	test('7. Out of range requests - 416', done => {

		const options = {
			headers: { 'Range': 'bytes=0-14' },
		} 

		http.get( 'http://127.0.0.1:9000', options, response => {

			expect( response.statusCode ).toBe( 416 );
			expect( response.headers[ 'content-range' ] ).toBe( 'bytes */12' );
			done();

		})

	});

	test('8. Testing maxFileSize', done => {

		http.get( 'http://127.0.0.1:9003', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'File served through "fs.createReadStream"' );
				done();

			})

		})

	});

	test('9. Calling external middlewares"', done => {

		http.get( 'http://127.0.0.1:9004/this_doesnt_exist.txt', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'External middlewares are being called on 404' );
				done();

			})

		})

	});

	test('10. Generate config when receiving just a string with a dir', done => {

		http.get( 'http://127.0.0.1:9001', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'Hello world!' );
				done();

			})

		})

	});

	test('11. Alias', done => {

		// Aliased request

		http.get( 'http://127.0.0.1:9005/components/component.html', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'Compiled! Yes!' );

			})

		})

		// Non-aliased request

		http.get( 'http://127.0.0.1:9005', response => {

			expect( response.statusCode ).toBe( 200 );

			let data = '';

			response.on( 'data', chunk => { data += chunk } )

			response.on( 'end', () => {

				expect( data ).toBe( 'Hello aliased world!' );
				done();

			})

		})

	});

	test('12. Serving memory generated files', async done => {

		const request = ( path = '', expected, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9000' + path, response => {

					expect( response.statusCode ).toBe( 200 );

					let data = '';

					response.on( 'data', chunk => { data += chunk } )

					response.on( 'end', () => {

						expect( data ).toBe( expected );

						promise_resolve();
						
						if ( end ) done();

					})

				})

			})

		}

		pulsa.memory( './test/public_html/memory_buffer.txt', Buffer.from( 'This buffer only exists in memory!' ) )
		pulsa.memory( './test/public_html/memory_file.txt', 'This file only exists in memory!' )

		await request( '/memory_buffer.txt', 'This buffer only exists in memory!' )
		await request( '/memory_file.txt', 'This file only exists in memory!' )

		pulsa.memory( './test/public_html/memory_file.txt', 'And we can update it!' )
		request( '/memory_file.txt', 'And we can update it!', true )

	});

	test('13. Ensuring that "clear" really clears the cache', async done => {

		const request = ( path = '', expected, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9006' + path, response => {

					expect( response.statusCode ).toBe( 200 );

					let data = '';

					response.on( 'data', chunk => { data += chunk } )

					response.on( 'end', () => {

						expect( data ).toBe( expected );

						promise_resolve();
						
						if ( end ) done();

					})

				})

			})

		}

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		fs.writeFileSync( 'test/public_changing/index.html', 'Unchanged index.' );
		fs.writeFileSync( 'test/public_changing/other.html', 'Unchanged other HTML file.' );

		await request( '/', 'Unchanged index.' )
		await request( '/other.html', 'Unchanged other HTML file.' )
		await request( '/inexistent/route', 'Unchanged index.' )

		fs.writeFileSync( 'test/public_changing/index.html', 'Changed index!' );
		fs.writeFileSync( 'test/public_changing/other.html', 'Changed the other HTML file as well!' );

		pulsa.clear( resolve( 'test/public_changing/index.html' ) )
		pulsa.clear( resolve( 'test/public_changing/other.html' ) )

		await request( '/', 'Changed index!' )
		await request( '/other.html', 'Changed the other HTML file as well!' )
		await request( '/inexistent/route', 'Changed index!' )

		// Undo the changes
		fs.writeFileSync( 'test/public_changing/index.html', 'Unchanged index.' );
		fs.writeFileSync( 'test/public_changing/other.html', 'Unchanged other HTML file.' );

		done();

	});

	test('14. SPA working with on-memory indexes!', async done => {

		const request = ( path = '', expected, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9007' + path, response => {

					expect( response.statusCode ).toBe( 200 );

					let data = '';

					response.on( 'data', chunk => { data += chunk } )

					response.on( 'end', () => {

						expect( data ).toBe( expected );

						promise_resolve();
						
						if ( end ) done();

					})

				})

			})

		}

		pulsa.memory( './test/public_spa_memory/index.html', 'This SPA is serverd from memory!' )

		await request( '/', 'This SPA is serverd from memory!' )
		await request( '/inexistent/route', 'This SPA is serverd from memory!' )

		pulsa.memory( './test/public_spa_memory/index.html', 'This SPA is UPDATED from memory!' )

		await request( '/', 'This SPA is UPDATED from memory!' )
		request( '/inexistent/route', 'This SPA is UPDATED from memory!', true )

	});

	test('15. Memory files overwriting 404 files', async done => {

		const request = ( path = '', expected, non_existent = false, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9000' + path, response => {

					expect( response.statusCode ).toBe( non_existent ? 404 : 200 );

					if ( non_existent ) {

						promise_resolve();

						if ( end ) done();

					} else {

						let data = '';

						response.on( 'data', chunk => { data += chunk } )

						response.on( 'end', () => {

							expect( data ).toBe( expected );

							promise_resolve();
							
							if ( end ) done();

						})

					}

				})

			})

		}

		await request( '/non_existen_to_be_overwrited.txt', null, true )

		pulsa.memory( './test/public_html/non_existen_to_be_overwrited.txt', 'Now this non-existent file exists! No more 404s!' )

		request( '/non_existen_to_be_overwrited.txt', 'Now this non-existent file exists! No more 404s!', false, true )

	});

	test('16. Memory files overwriting 404 index-files', async done => {

		const request = ( path = '', expected, non_existent = false, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9008' + path, response => {

					expect( response.statusCode ).toBe( non_existent ? 404 : 200 );

					if ( non_existent ) {

						promise_resolve();

						if ( end ) done();

					} else {

						let data = '';

						response.on( 'data', chunk => { data += chunk } )

						response.on( 'end', () => {

							expect( data ).toBe( expected );

							promise_resolve();
							
							if ( end ) done();

						})

					}

				})

			})

		}

		await request( '/', null, true )
		await request( '/dir', null, true )

		pulsa.memory( './test/public_memory/index.html', 'Now this non-existent INDEX exists! No more 404s!' )
		pulsa.memory( './test/public_memory/dir/index.html', 'Now this non-existent INDEX inside "dir" exists! No more 404s!' )

		await request( '/', 'Now this non-existent INDEX exists! No more 404s!', false )
		request( '/dir', 'Now this non-existent INDEX inside "dir" exists! No more 404s!', false, true )

	});

	test('17. Memory files overwriting existing files from disk', async done => {

		const request = ( path = '', expected, end = false ) => {

			return new Promise ( promise_resolve => {

				http.get( 'http://127.0.0.1:9009' + path, response => {

					expect( response.statusCode ).toBe( 200 );

					let data = '';

					response.on( 'data', chunk => { data += chunk } )

					response.on( 'end', () => {

						expect( data ).toBe( expected );

						promise_resolve();
						
						if ( end ) done();

					})

				})

			})

		}

		await request( '/', 'This content needs to be overwrited!' )
		await request( '/other.html', 'This other content needs to be overwrited!' )

		pulsa.memory( './test/public_overwriting/index.html', 'Index overwrited, even existing. Nicely done!' )
		pulsa.memory( './test/public_overwriting/other.html', 'Other overwrited, even existing. Nicely done!' )

		await request( '/', 'Index overwrited, even existing. Nicely done!' )
		await request( '/other.html', 'Other overwrited, even existing. Nicely done!' )

		done();

	});

	test('Final: Clearing caches test', done => {

		// With a path

		pulsa.clear( resolve( './test/public_html/index.html' ) )
		pulsa.clear( resolve( './test/public_spa/without_index/' ) + '/' )
		pulsa.clear( resolve( './test/public_spa/spa/without_index/' ) )

		// Without a path (ignored)
		pulsa.clear()

		// Done
		done();

	});

});

src/index.js
import fs 						from 'fs';
import mime						from 'mime/lite';
import { join, resolve, sep }	from 'path';
import parser					from '@polka/url';
import { Readable } 			from 'stream';


// Shortcuts
const readStream 	= fs.createReadStream;
const read 			= fs.readFileSync;

// Pulsa internal data
const caching 	= {};
const map 		= {};
let   instances = 0;
const instance 	= {};




const ensure = path => {
	if ( !map[ path ] ) map[ path ] = { reverses: {}, ranges: {} };
}

const reverse = ( path, run, pathname ) => {

	if ( !map[ path ].reverses[ run.id ] ) map[ path ].reverses[ run.id ] = {};

	map[ path ].reverses[ run.id ][ pathname ] = true;

};

const update_instances = function ( path, response, remove = false ) {

	if ( !map[ path ] ) return;

	const ref = map[ path ];

	for ( let id in ref.reverses ) {
		
		const paths 	= ref.reverses[ id ];
		const run 		= instance[ id ];

		for ( let pathname in paths ) {
			
			if ( remove )
				delete run.responses[ pathname ];
			else {
				run.responses[ pathname ] = response;
			}
		}

	}

}

const check = function ( path ) {

	ensure( path );

	if ( map[ path ].stats )
		return true;

	try {
		map[ path ].stats = fs.statSync( path );
		return true;
	}
	catch ( e ) {
		return false;
	}

}

const check_index = function ( path ) {

	let sufix = '';

	if ( !check( path ) || ( map[ path ].stats.isDirectory() && ( !check( path + ( sufix = sep + 'index.html' ) ) || map[ path + sufix ].stats.isDirectory() ) ) )
		return { stop: true };

	return { stop: false, sufix };

}

const cache_stream = function ( path, run, buffer, ref ) {

	if ( run && ref.stats.size > run.maxFileSize )
		return ref.stream = ( res, opts ) => readStream( path, opts ).pipe( res );

	ref.data = buffer || read( path );

	return ref.stream = ( res, opts ) => {

		const s = new Readable;
		s.push( opts.end ? ref.data.slice( opts.start, opts.end + 1 ) : ref.data );
		s.push( null );

		s.pipe( res );

	}

}

const cache_range = function ( range, headers, ref ) {

	const opts = {}, size = ref.stats.size;

	let [ x, y ] = range.replace( 'bytes=', '' ).split( '-' );
	let end = opts.end = parseInt( y, 10 ) || ( size - 1 );
	let start = opts.start = parseInt( x, 10 ) || 0;

	if ( start >= size || end >= size ) {
		
		const content_range = `bytes */${size}`;

		return function ( res ) {

			res.setHeader( 'Content-Range', content_range );
			res.statusCode = 416;
			return res.end();

		}
	}

	const range_headers = {
		'Content-Range': 	`bytes ${start}-${end}/${size}`,
		'Content-Length': 	( end - start + 1 ),
		'Accept-Ranges': 	'bytes',
		'Content-Type': 	headers[ 'Content-Type' ],
		'Last-Modified': 	headers[ 'Last-Modified' ]
	}

	return function ( res ) {

		res.writeHead( 206, range_headers );
		ref.stream( res, opts );

	}

}

const response = function ( path, sufix, ref ) {

	const sufix_ref = sufix ? map[ path + sufix ] : ref;

	ref.headers = {
		'Content-Length': sufix_ref.stats.size,
		'Content-Type': mime.getType( path ) + ( ref.charset ? '; charset=' + ref.charset : '' ),
		'Last-Modified': sufix_ref.stats.mtime.toUTCString(),
	};

	return ref.response = sufix_ref.response = function ( req, res ) {

		if ( req.headers.range ) {
			const range = req.headers.range;
			( ref.ranges[ range ] || ( ref.ranges[ range ] = cache_range( range, ref.headers, sufix_ref ) ) )( res );

		}
		else {
			res.writeHead( 200, ref.headers );
			sufix_ref.stream( res, {} );
		}

	}

}

const alias = function ( path, run ) {

	if ( !run.alias || typeof run.alias !== 'object' )
		return path;

	for ( let key in run.alias ) {
		if ( path.startsWith( key ) )
			return path.replace( key, run.alias[ key ] );
	}

	return path;

}

const notFound = function ( req, res, next ) {
	return next ? next() : ( res.statusCode = 404, res.end() );
}

const clear = function ( path, recursive = true ) {

	if ( !path ) return;

	if ( path.length > 1 && path.endsWith( sep ) ) path = path.slice( 0, -1 );

	if ( recursive ) {

		if ( path.endsWith( sep + 'index.html' ) ) {

			clear( path.slice( 0, -11 ), false );

		} else if ( map[ path ] && map[ path ].stats && map[ path ].stats.isDirectory() )
			clear( path + sep + 'index.html', false );

	}

	update_instances( path, null, true );

	delete map[ path ];
	delete caching[ path ];

	return true;

}

const memory = function ( path, buffer ) {

	path = resolve( path );
	ensure( path );

	if ( !( buffer instanceof Buffer ) ) buffer = Buffer.from( buffer );

	const ref = map[ path ];

	ref.charset = 'utf-8';

	ref.stats = {
		size: 	buffer.length,
		mtime: 	new Date(),
		isDirectory: () => false
	}

	ref.stream = cache_stream( path, null, buffer, ref );
	ref.response = caching[ path ] = response( path, '', ref );

	update_instances( path, ref.response );

	// Directories behind the "index.html"
	if ( path.endsWith( sep + 'index.html' ) )
		clear( path.slice( 0, -11 ), false );

	return true;

}

const cache = async function ( pathname, run ) {

	let path = join( run.dir, alias( pathname, run ) );

	if ( path.endsWith( sep ) ) path = path.slice( 0, -1 );

	if ( caching[ path ] ) return caching[ path ];

	ensure( path );
	reverse( path, run, pathname );

	const ref 				= map[ path ];
	const { stop, sufix } 	= check_index( path );

	if ( stop ) {
		
		let spa_path = join( run.dir, ( run.base.endsWith( '/' ) ? run.base.slice( 0, -1 ) : run.base ) );

		if ( run.spa ) {
			reverse( spa_path, run, pathname );
			return ref.response = caching[ spa_path ];
		}
		else
			return ref.response = notFound;
	}
	else reverse( path + sufix, run, pathname );

	const sufix_ref = map[ path + sufix ];

	ref.stream = sufix_ref.stream = ref.stream || sufix_ref.stream || cache_stream( path + sufix, run, null, ref );
	ref.response = sufix_ref.response = ref.response || sufix_ref.response || response( path, sufix, ref );

	return caching[ path ] = caching[ path + sufix ] = ref.response;

}

const serve = function ( config ) {

	if ( typeof config === 'string' )
		config = { dir: config };

	const run = Object.assign({
		base: 			'/',
		dir: 			'.',
		maxFileSize: 	1048576, // 1 MB
		spa: 			false
	}, config );

	run.id 				= ++instances;
	instance[ run.id ] 	= run;
	run.dir 			= resolve( run.dir );
	run.responses 		= {};
	run.base 			= alias( run.base, run );

	cache( run.base, run );

	return async function ( req, res, next ) {
	
		const pathname = decodeURIComponent( req.path || req.pathname || parser( req ).pathname );

		return ( run.responses[ pathname ] || ( run.responses[ pathname ] = await cache( pathname, run ) ) )( req, res, next );

	}
}

export { serve, memory, clear };